docs/source/developer/performance.rst

```diff
@@ -0,0 +1,62 @@
+Performance tricks
+==================
+
+Notebook documents
+------------------
+
+Performance analysis of notebook documents in JupyterLab point out to two main bottlenecks:
+editors style computation and rendered markdown with lots of math expressions. The best
+technique to alleviate those problems are to attach to the DOM only the cells in the viewport.
+However code cell outputs bring a huge constraint that their internal state is within the
+DOM node (e.g. the zoom level on a map view). So the outputs cannot be detached once displayed
+With those considerations in mind, here is the algorithm coded for the notebook windowing.
+
+When cell widgets are instantiated, their children are not created (i.e. the editor, the
+outputs,…) and they are not attached to the DOM. The view is updated on scroll events following:
+
+1. Get the scroll position in the notebook
+2. Get the cells range to be displayed
+3. Attach the cells in viewport to the notebook
+   a. Before attaching the cell, if it was never attached, instantiate its children.
+   b. For code cells previously attach, change its visibility and attach all children except the output area.
+   c. Add the cell widget to a resize observer that trigger and viewport update
+4. Detach cells that left the viewport
+   a. Remove cell widget from resize observer.
+   b. For code cells, the output area is not detached but hidden. All other children are detached.
+
+Due to the code cells being kept in the DOM, we need to update the data attribute
+``data-windowed-list-index`` when the cells list changes. This is required to attach at the
+correct position the cell when the viewport changes.
+
+The list windowing algorithm is coded in a dedicated widget ``WindowedList``. The notebook
+extends it and uses a specialized layout to deal with code cells, ``NotebookWindowedLayout``.
+The code cell layout is also customized ``CodeCellLayout`` to keep the output area attached
+but not the other children.
+
+.. warning::
+
+    The current implementation does not handle *iframe* well. In Chrome, the iframe states are
+    reset every time they leave and renter the viewport. In Firefox, it does not happen when
```

Comment:
This is a tricky case. My plan would be first to get this version in and ship it as disabled by default due to that limitation.

The follow-up would be:
- Figure out how to deal with iframes  
  I'ld like to avoid the trick of Cocalc to have them positioned absolutely so that they remain in a DOM node not impacted by the DOM reorganization due to windowing.
- Figure out how to deal with outputs that contains Javascript and/or CSS that may not be in the initial viewport. And therefore the features and the look and feel is not proper.

Reply:
I use iframes a lot in my educational notebooks, so would definitely like to have a fix for this.

Two thoughts, fiddling with a notebook, and applying styles to the `.jp-Cell-outputWrapper`:

```css
z-index: -1;
position: absolute;
```
The Cocalc trick you mentioned.

```css
height:0;
overflow: auto;
```
This seems to hide the frame as well and, on Chrome at least, preserves the scroll state.